### PR TITLE
Fix/swizzling and deprecated apis

### PR DESCRIPTION
**Pull request by clpolet**
Deprecated APIs won't be swizzled if the original app delegate only implement the new API or don't don't implement at all.

### CODE REVIEW

File: MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m [hunk 1]
```diff
@@ -79,13 +86,23 @@ + (void)setDelegates:(NSHashTable<id<MSAppDelegate>> *)delegates {
...
-        @[ NSStringFromSelector(@selector(application:didReceiveRemoteNotification:fetchCompletionHandler:)) ];
+    _selectorsNotToOverride = @[ kMSDidReceiveRemoteNotificationFetchHandler ];
 #endif
   }
   return _selectorsNotToOverride;
 }
 
++ (NSDictionary<NSString *, NSString *> *)deprecatedSelectors {
+  if (!_deprecatedSelectors) {
+#if !TARGET_OS_OSX
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m]:
```
#if TARGET_OS_OSX
     _deprecatedSelectors = @{};
#else
     _deprecatedSelectors = @{kMSOpenURLOptions : kMSOpenURLSourceApplicationAnnotation};
#endif
```
`!` is unnecessary.

File: MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m [hunk 2]
```diff
@@ -179,17 +198,40 @@ + (IMP)swizzleOriginalSelector:(SEL)originalSelector
...
-     * add the method to the original class and associate it with the custom implementation.
-     */
-    Method customMethod = class_getInstanceMethod(self, customSelector);
-    methodAdded = class_addMethod(originalClass, originalSelector, customImp, method_getTypeEncoding(customMethod));
+    // Check for deprecation.
+    NSString *deprecatedSelectorStr = self.deprecatedSelectors[originalSelectorStr];
+    if (deprecatedSelectorStr &&
+        [originalClass instancesRespondToSelector:NSSelectorFromString(deprecatedSelectorStr)]) {
+
+      // An implementation for the deprecated selector exists. Don't add the new method, it might eclipse the original
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m]: Multiline comment. Use `/* */`

File: MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m [hunk 3]
```diff
@@ -179,17 +198,40 @@ + (IMP)swizzleOriginalSelector:(SEL)originalSelector
...
+      // implementation.
+      warningMsg = [NSString
+          stringWithFormat:
+              @"No implementation found for this selector, though an implementation of its deprecated API '%@' exists.",
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m]: Why are you saying `No implementation found for this selector` here? The above comment says it found an implementation of the deprecated selector. Am I missing something?

File: MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m (hunk 1 above)
Comment (clpolet) [reply MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m]: Will do.

File: MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m (hunk 2 above)
Comment (clpolet) [reply MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m]: Good catch, it got multilined by the formatter.

File: MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m (hunk 3 above)
Comment (clpolet) [reply MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m]: An implementation for the deprecated selector has been found but no implementation were found for the current API which is the new API supposed to replace the deprecated one ;).

Comment (clpolet) [reply MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m]: I'll rephrase the comment.

File: MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m [hunk 4]
```diff
@@ -179,17 +198,40 @@ + (IMP)swizzleOriginalSelector:(SEL)originalSelector
      * depend on the SDK return value for its own logic so customers already have to call the SDK API
      * in their implementation which makes swizzling useless.
      */
-    if (![self.selectorsNotToOverride containsObject:originalSelectorString]) {
+    if (![self.selectorsNotToOverride containsObject:originalSelectorStr]) {
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m]: It's just 3 more letters, why not write them? ;)

File: MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m (hunk 3 above)
Comment (TroubleMakerBen) [reply MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m]: Maybe we can be more explicit here and add a suggestion, e.g. "Please implement it's successor" or something like that.

File: MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarderPrivate.h [hunk 5]
```diff
@@ -22,6 +22,11 @@ NS_ASSUME_NONNULL_BEGIN
 @property(nonatomic, class, readonly) NSArray<NSString *> *selectorsNotToOverride;
 
 /**
+ * Dictionary of deprecated original selectors indexed by their new equivalent.
+ */
+@property(nonatomic, class, readonly) NSDictionary<NSString *, NSString *> *deprecatedSelectors;
+
+/**
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarderPrivate.h]: I like this!

File: MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m (hunk 4 above)
Comment (clpolet) [reply MobileCenter/MobileCenter/Internals/AppDelegate/MSAppDelegateForwarder.m]: I usually shorten local variables name when they start to get big, I felt it was the case here. There are couple of other examples in the this page.

**Pull request merged**